### PR TITLE
(feat) O3-4143: useLocations should support searching by tag or query…

### PR DESCRIPTION
**Pull request by mogoodrich**
… string

# Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. Ensure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing#contributing-guidelines) label (such as `feat`, `fix`, or `chore`, among others). See existing PR titles for inspiration.

## For changes to apps

- [ ] My work conforms to the [**O3 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://om.rs/o3ui).

## If applicable

- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary

see: https://openmrs.atlassian.net/browse/O3-4143

## Related Issue

https://openmrs.atlassian.net/browse/O3-4143

### CODE REVIEW

**Comment by mogoodrich** [comment]
I wasn't seeing any existing tests around these functions to update?  What is our policy for tests around stuff like this?

**Comment by mogoodrich** [comment]
@ibacher  @denniskigen  just a ping... thoughts on this one?  I need this to be able to add the custom behavior to filter visit locations by the "Visit Location" tag while still maintaining the ability to search locations by string.

File: packages/framework/esm-api/src/shared-api-objects/location.ts [hunk 1]
```diff
@@ -11,8 +11,16 @@ export function toLocationObject(openmrsRestForm: any): Location {
...
+  tagUuidOrName: string | null = null,
+  query: string | null = null,
+): Observable<Array<Location>> {
+  const url =
+    `${restBaseUrl}/location` +
+    (tagUuidOrName || query ? '?' : '') +
+    (tagUuidOrName ? 'tag=' + tagUuidOrName : '') +
+    (tagUuidOrName && query ? '&' : '') +
+    (query ? 'q=' + query : '');
   return openmrsObservableFetch<any>(url)
```
Comment (denniskigen) [inline packages/framework/esm-api/src/shared-api-objects/location.ts:24]:
```suggestion
  const params = new URLSearchParams();
  if (tagUuidOrName) {
    params.set('tag', tagUuidOrName);
  }
  if (query) {
    params.set('q', query);
  }
  const queryString = params.toString();
  const url = `${restBaseUrl}/location${queryString ? '?' + queryString : ''}`;

  return openmrsObservableFetch<{ results: Array<Location> }>(url)
```

Would something like this work as well? The current URL construction could produce an invalid URL when `tagUuidOrName` is null but query is provided. In this case, the URL would include an empty string where the tag= parameter should be, potentially resulting in malformed URLs like:

`/location?&q=searchterm`

Also, this annotates the return type of the locations response correctly.

**Review by denniskigen — APPROVED** [review]
Thanks, @mogoodrich. Mostly LGTM save for one concern about the URL construction.

File: packages/framework/esm-api/src/shared-api-objects/location.ts (hunk 1 above)
Comment (mogoodrich) [reply packages/framework/esm-api/src/shared-api-objects/location.ts:24]: Perfect!  Thanks so much @denniskigen ... I *think* I handled the null case, but this is *much* cleaner... I will commit your change, test, and then merge if all is well...

**Pull request merged**